### PR TITLE
INT-4219: Add @UseSpelInvoker

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-4219

Needs (a lot) more work, tests, docs.

Just want to get buy-in for the approach.

Also see note about resolving placeholders for compiler mode.

### CODE REVIEW

**Review by artembilan — CHANGES_REQUESTED** [review]
I we're good to go with this approach and that doesn't look so ugly if you don't like to use `InvocableMessageHandler`.

I don't see how you apply that `compilerMode`, but I guess that willl in upcoming commits.

File: spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java [hunk 1]
```diff
@@ -243,6 +243,10 @@ protected String getInputChannelAttribute() {
 		return INPUT_CHANNEL_ATTRIBUTE;
 	}
 
+	protected ConfigurableListableBeanFactory getBeanFactory() {
+		return this.beanFactory;
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java]: Yeah - this is some left-over dross from when I was trying to use the BF too early.

File: spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ * <p>Applying this annotation to those methods will cause SpEL to be used for the
+ * invocation. An optional compilerMode property (aliased to value) is also provided.
+ *
+ * @author Gary Russell
+ */
+@Target({ ElementType.TYPE, ElementType.METHOD, ElementType.ANNOTATION_TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+@Documented
+public @interface UseSpelInvoker {
```
Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java]: I don't think so; we will need that for XML config (either a new attribute or they can autowire the handler and set the flag).

Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java]: ??? Even for XML config the POJO method can be marked with the `@UseSpelInvoker`.
We don't prohibit to use `@ServiceActivator` and similar when we parse XML config.
Finally everything ends up in the `MessagingMethodInvokerHelper` which just reads a `Method` for particular options and generates SpEL and `InvocableHandlerMethod` for runtime.

Comment (garyrussell) [inline spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java]: The problem is they might want to invoke a class/method they don't own and can't annotate.

**Comment by garyrussell** [comment]
Next iteration.

**Comment by artembilan** [comment]
Well, I think the progress is good here.
You can go ahead what you want to do here else.

And I see your point about an external class to use from XML.
There is always a workaround like local wrapper around that remote method.
My point just don't spend time for XML effort.
At least until strong requirements.

**Comment by garyrussell** [comment]
>My point just don't spend time for XML effort.

Agreed.

**Comment by artembilan** [comment]
FYI: there are some failures according Travis report.

**Comment by artembilan** [comment]
Well, looks like we need only docs on the matter already!
And resolving conflicts after rebase.
Do you have anything else in mind to do for this task ?

**Comment by garyrussell** [comment]
PR soon (tests and docs) - doing a `testAll` right now 😄

File: spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+	/**
+	 * Specify that the annotated method (or methods in the annotated class) will be
+	 * invoked using SpEL instead of an
+	 * {@link org.springframework.messaging.handler.invocation.InvocableHandlerMethod}
+	 * with the specified compilerMode. If left empty, the default runtime compiler
+	 * mode will be used. Must evaluate to a String containing a valide compiler mode.
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java]: Typo: valid

Will fix on merge

File: spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	 */
+	@AliasFor("compilerMode")
+	String value() default "";
+
+	/**
+	 * Specify that the annotated method (or methods in the annotated class) will be
+	 * invoked using SpEL instead of an
+	 * {@link org.springframework.messaging.handler.invocation.InvocableHandlerMethod}
+	 * with the specified compilerMode. If left empty, the default runtime compiler
+	 * mode will be used. Must evaluate to a String containing a valide compiler mode.
```
Comment (artembilan) [inline spring-integration-core/src/main/java/org/springframework/integration/annotation/UseSpelInvoker.java]: DITO.
Copy/paste artefact 😄

**Review by artembilan — APPROVED** [review]
A couple (in fact one) typos.
Merging...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/a50d866f81888f27c56f7298d946b786c52e9281

**Pull request closed without merging**